collector/mongod/oplog_status.go

```diff
@@ -1,102 +1,116 @@
-            Name:      "size_mb",
-            Help:      "Size of oplog in megabytes",
-    })
+	oplogStatusCount = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace:	Namespace,
+		Subsystem:	"replset_oplog",
+		Name:		"items_total",
+		Help:		"The total number of changes in the oplog",
+	})
+	oplogStatusHeadTimestamp = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace:	Namespace,
+		Subsystem:	"replset_oplog",
+		Name:		"head_timestamp",
+		Help:		"The timestamp of the newest change in the oplog",
+	})
+	oplogStatusTailTimestamp = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace:	Namespace,
+		Subsystem:	"replset_oplog",
+		Name:		"tail_timestamp",
+		Help:		"The timestamp of the oldest change in the oplog",
+	})
+	oplogStatusSizeBytes = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace:	Namespace,
+		Subsystem:	"replset_oplog",
+		Name:		"size_bytes",
+		Help:		"Size of oplog in bytes",
+	}, []string{"type"})
 )
 
-func GetCollectionSizeMB(db string, collection string, session *mgo.Session) (float64) {
-    var collStats map[string]interface{}
-    err := session.DB(db).Run(bson.D{{"collStats", collection }}, &collStats)
-    if err != nil {
-        glog.Error("Error getting collection stats!")
-    }
-
-    var result float64 = -1
-    if collStats["size"] != nil {
-        size := collStats["size"].(int)
-        result = float64(size)/1024/1024
-    }
-
-    return result
+type OplogCollectionStats struct {
+	Count		float64	`bson:"count"`
+	Size		float64	`bson:"size"`
+	StorageSize	float64 `bson:"storageSize"`
 }
 
-func GetOplogSizeMB(session *mgo.Session) (float64) {
-    return GetCollectionSizeMB("local", "oplog.rs", session)
+type OplogTimestamps struct {
+	Tail	float64
+	Head	float64
 }
 
-func ParseBsonMongoTsToUnix(timestamp bson.MongoTimestamp) (int64) {
-    return int64(timestamp >> 32)
+type OplogStatus struct {
+	OplogTimestamps	*OplogTimestamps
+	CollectionStats	*OplogCollectionStats
 }
 
-type OplogStatsData struct {
-    MinTime	bson.MongoTimestamp	`bson:"min"`
-    MaxTime	bson.MongoTimestamp	`bson:"max"`
+// there's gotta be a better way to do this, but it works for now :/
+func BsonMongoTimestampToUnix(timestamp bson.MongoTimestamp) float64 {
+	return float64(timestamp >> 32)
 }
 
-func GetOplogLengthSecs(session *mgo.Session) (float64, float64) {
-    results := &OplogStatsData{}
-    group := bson.M{ "_id" : 1, "min" : bson.M{ "$min" : "$ts" }, "max" : bson.M{ "$max" : "$ts" } }
-    err := session.DB("local").C("oplog.rs").Pipe([]bson.M{{ "$group" : group  }}).One(&results)
-    if err != nil {
-        glog.Error("Could not get the oplog time min/max!")
-        return -1, -1
-    }
-
-    minTime := ParseBsonMongoTsToUnix(results.MinTime)
-    maxTime := ParseBsonMongoTsToUnix(results.MaxTime)
-
-    now := time.Now().Unix()
-    lengthSeconds := maxTime - minTime
-    lengthSecondsNow := now - minTime
-
-    return float64(lengthSeconds), float64(lengthSecondsNow)
```

Comment:
The previous code had 2 x values for oplog time: "length" and "length to now", which are 2 x float64 values. This new code now returns "head" and "tail" which is also 2 x float64 values that can provide more flexibility in graphing in the future. I feel this is a more optimal approach but this can be rolled back if we must, however this will require updates and retesting of the grafana_mongodb_dashboards project also.